### ISSUE TITLE
Error mounting ghostfs in agent initialization

### ISSUE THREAD

**Issue by EmilyPetersonM** [google/ghost-userspace#18]
I'm attempting to run the fifo_centralized_agent (or any agent) after I have gotten the target to build successfully. However, upon running `bazel run fifo_centralized_agent`, I encounter an error:
```root@localhost:~/ghosty/ghost-userspace# bazel run fifo_centralized_agent
INFO: Analyzed target //:fifo_centralized_agent (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //:fifo_centralized_agent up-to-date:
  bazel-bin/fifo_centralized_agent
INFO: Elapsed time: 15.846s, Critical Path: 12.69s
INFO: 3 processes: 1 internal, 2 processwrapper-sandbox.
INFO: Build completed successfully, 3 total actions
INFO: Build completed successfully, 3 total actions
lib/ghost.cc:150(19449) CHECK FAILED: (*__errno_location ()) == 16 [2 != 16]
errno: 2 [No such file or directory]
PID 19449 Backtrace:
[0] 0xaaaabd3f521c : ghost::Exit()
[1] 0xaaaabd3eace0 : ghost::Ghost::MountGhostfs()
[2] 0xaaaabd3ead3c : ghost::Ghost::GetSupportedVersions()
[3] 0xaaaabd392db8 : ghost::Ghost::CheckVersion()
[4] 0xaaaabd392674 : __static_initialization_and_destruction_0()
[5] 0xaaaabd3926a4 : _GLOBAL__sub_I_agent.cc
[6] 0xaaaabd4a4108 : __libc_csu_init
[7] 0xffff94507cf8 : __libc_start_main
```
The ENOENT error indicates a failure to mount ghostfs because the mount point "/sys/fs/ghost" is missing. My output from executing `mount -t ghost ghost /sys/fs/ghost` is `mount: /sys/fs/ghost: mount point does not exist.`

Do you have any idea what might be causing this error, or what is wrong with my installation of ghOSt? I should be running the ghOSt kernel version 62, but it is notable that the machine everything was built for and running on is arm64 (which has required some custom tweaks to build but not many) and it's also an embedded board. Thanks in advance for any help!

**Comment by jackhumphries** [comment]
Hi,

Are you sure that you installed and booted the ghOSt kernel? If the mount point is missing, it seems as though the ghOSt kernel is not running on your machine. Can you send me the output of `uname -a`?

Thanks,
Jack

**Comment by jackhumphries** [comment]
As a follow-up note, did you add `CONFIG_SCHED_CLASS_GHOST=y` to your `.config`?

**Comment by EmilyPetersonM** [comment]
My output for `uname -a` is `Linux localhost 5.11.0-g5da05ec77890 #3 SMP PREEMPT Thu May 5 00:00:04 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux`.

I am missing the `CONFIG_SCHED_CLASS_GHOST` line from my `.config`, which for some reason I thought was x86 specific. I am now trying to rebuild the kernel with this line added and see if it behaves differently.

**Comment by EmilyPetersonM** [comment]
As an update, I was able to add `CONFIG_SCHED_CLASS_GHOST=y` to the configuration but only by adding it into the kernel make command because the line kept getting removed automatically when I added it manually into the `.config` file. Afterwards, it looks like the ghost portions of the kernel finally built for the first time because I immediately got a slew of compile errors from files such as kernel/sched/ghost.c. I am guessing the compile errors have to do with my version of gcc or clang (which should be maximally up-to-date). But for example, the first compile error is:
```
In file included from ./include/linux/rculist.h:11,
                 from ./include/linux/dcache.h:8,
                 from ./include/linux/fs.h:8,
                 from ./include/uapi/linux/aio_abi.h:31,
                 from ./include/linux/syscalls.h:73,
                 from kernel/sched/ghost.c:14:
kernel/sched/ghost.c: In function 'check_same_enclave':
kernel/sched/ghost.c:107:30: error: 'enclave' undeclared (first use in this function)
  107 |  x = rcu_dereference(per_cpu(enclave, cpu_x));
      |                              ^~~~~~~
```
and the second one is:
```
In file included from ./include/linux/spinlock.h:59,
                 from ./include/linux/wait.h:9,
                 from ./include/linux/wait_bit.h:8,
                 from ./include/linux/fs.h:6,
                 from ./include/uapi/linux/aio_abi.h:31,
                 from ./include/linux/syscalls.h:73,
                 from kernel/sched/ghost.c:14:
kernel/sched/ghost.c: In function 'enclave_is_dying':
kernel/sched/ghost.c:120:24: error: dereferencing pointer to incomplete type 'struct ghost_enclave'
  120 |  lockdep_assert_held(&e->lock);
      |                        ^~
```
and the third:
```
kernel/sched/ghost.c: In function 'ghost_bpf_pnt':
kernel/sched/ghost.c:194:5: error: 'struct rq' has no member named 'ghost'
  194 |   rq->ghost.dont_idle_once = true;
      |     ^~
```
:( I'm planning to look at these errors more tomorrow. But you are totally right that ghOST was not running and not built into the kernel image! This is a very helpful launching point.

**Comment by jackhumphries** [comment]
Hi Emily,

I took a closer look and the reason why `CONFIG_SCHED_CLASS_GHOST=y` is being removed from your .config is because of ghOSt's entry in `init/Kconfig`:
```
config SCHED_CLASS_GHOST
        bool "ghOSt scheduling class"
        depends on X86_64
        default n
        help 
          ghOSt is a scheduling class with policy decisions delegated
          to a user-space agent.
```
If you remove `depends on X86_64` from that entry (because you are compiling for ARM), then `CONFIG_SCHED_CLASS_GHOST=y` should stay in your `.config` file if you add it again. I would not add `CONFIG_SCHED_CLASS_GHOST` to the `make` command.

Please try this and let me know if the compile succeeds and if you are able to run ghOSt on your ARM machine. We haven't tried running ghOSt on ARM yet ourselves, but I know others outside of Google have run ghOSt on ARM machines, so this should hopefully fix the problem.

**Comment by EmilyPetersonM** [comment]
Thanks so much for your help!

I noticed the `depends on X86_64` line and removed it from the Kconfig, then rebuilt the kernel. As you mention, having that config appropriately set took care of the compile errors I was seeing. I also had to go back and set `CONFIG_BPF_SYSCALL=y` (and rebuild again) which was missing from my `.config` and needed by ghOSt. After that, the scheduler agent ran without issue!

For the record here, to get up and running on ARM I also needed to replace some lines of code in the userspace library using the `pause` instruction that is not supported on ARM. Specifically, I replaced all lines `asm volatile("pause");` with `asm volatile("yield");` If I remember correctly, these were the only changes I had to make when building both the kernel and userspace component in order to get to the point of running a ghOSt agent on my aarch64 machine.

Thanks again for the help, and I'm excited to start using ghOSt for some scheduler prototyping and experiments!
-Emily Peterson
